compiler/rustc_typeck/src/check/demand.rs

```diff
@@ -145,12 +145,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let expr_ty = self.resolve_vars_with_obligations(checked_ty);
         let mut err = self.report_mismatched_types(&cause, expected, expr_ty, e);
 
-        if self.is_assign_to_bool(expr, expected) {
-            // Error reported in `check_assign` so avoid emitting error again.
```

Comment:
So, I can't find a `check_assign`. There are couple other places (`coercion.rs` and `expr.rs`) that we also reference this. If you remove those special cases too, what happens?

I'm really not sure why this is special cased.

Reply:
The comments are out of date. I took a look at the git blame, and I found [this commit](https://github.com/rust-lang/rust/commit/7945eff0805d91d3c0ca10df872bcabdd88aa4b6) that introduced the special check for `is_assign_to_bool`. 

The above commit also introduces `check_assign` (later re-named to `check_expr_assign`, as seen [here](https://github.com/rust-lang/rust/commit/af800c7873b7dfa28a8583e2e89d4c573d525bb5)). We can see that it makes a call to `check_expr_coercable_to_type` and then emits a dialog if the user likely meant to use `==` instead of `=`. So at the time of the commit, the check for `is_assign_to_bool` was required to prevent unnecessary messages from the call to `check_expr_coercable_to_type`

I am unsure of what happens if we remove all the special cases. I will investigate.

Reply:
Interestingly, no UI tests fail after I remove the checks to `is_assign_to_bool`.

Reply:
Let's just remove all checks. At worst, we just error more. (If we find a test case where there are duplicate errors, we can revisit then)